x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/Coalesce.java

```diff
@@ -194,70 +201,16 @@ public boolean foldable() {
 
     @Override
     public ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
-        List<ExpressionEvaluator.Factory> childEvaluators = children().stream().map(toEvaluator::apply).toList();
-        return new ExpressionEvaluator.Factory() {
-            @Override
-            public ExpressionEvaluator get(DriverContext context) {
-                return new CoalesceEvaluator(
-                    context,
-                    PlannerUtils.toElementType(dataType()),
-                    childEvaluators.stream().map(x -> x.get(context)).toList()
-                );
-            }
-
-            @Override
-            public String toString() {
-                return "CoalesceEvaluator[values=" + childEvaluators + ']';
-            }
+        return switch (dataType()) {
+            case BOOLEAN -> CoalesceBooleanEvaluator.toEvaluator(toEvaluator, children());
+            case DOUBLE, COUNTER_DOUBLE -> CoalesceDoubleEvaluator.toEvaluator(toEvaluator, children());
+            case INTEGER, COUNTER_INTEGER -> CoalesceIntEvaluator.toEvaluator(toEvaluator, children());
+            case LONG, DATE_NANOS, DATETIME, COUNTER_LONG, UNSIGNED_LONG -> CoalesceLongEvaluator.toEvaluator(toEvaluator, children());
+            case KEYWORD, TEXT, SEMANTIC_TEXT, CARTESIAN_POINT, CARTESIAN_SHAPE, GEO_POINT, GEO_SHAPE, IP, VERSION ->
+                CoalesceBytesRefEvaluator.toEvaluator(toEvaluator, children());
+            case NULL -> EvalOperator.CONSTANT_NULL_FACTORY;
+            case UNSUPPORTED, SHORT, BYTE, DATE_PERIOD, OBJECT, DOC_DATA_TYPE, SOURCE, TIME_DURATION, FLOAT, HALF_FLOAT, TSID_DATA_TYPE,
+                SCALED_FLOAT, PARTIAL_AGG -> throw new UnsupportedOperationException("can't be coalesced");
         };
```

Comment:
This `switch` will fail to compile if you make a new type and don't include it. We intend for COALESCE to work for all types. If you are building a new type then you can stub it out as you go like you'll do with a few other evaluators.